lfs.c

```diff
@@ -5233,40 +5233,63 @@ static int lfs_fs_gc_(lfs_t *lfs) {
 #endif
 
 #ifndef LFS_READONLY
+#ifdef LFS_SHRINKIFCHEAP
+static int lfs_shrink_checkblock(void * data, lfs_block_t block) {
+    lfs_size_t threshold = *((lfs_size_t *) data);
+    if (block >= threshold) {
+        return LFS_ERR_NOTEMPTY;
+    }
+    return 0;
+}
+#endif
+
 static int lfs_fs_grow_(lfs_t *lfs, lfs_size_t block_count) {
+    int err;
+
+    if (block_count == lfs->block_count) {
+        return 0;
+    }
+
+    
+#ifndef LFS_SHRINKIFCHEAP
     // shrinking is not supported
     LFS_ASSERT(block_count >= lfs->block_count);
+#endif
+#ifdef LFS_SHRINKIFCHEAP
+    lfs_block_t threshold = block_count;
+    err = lfs_fs_traverse_(lfs, lfs_shrink_checkblock, &threshold, true);
```

Comment:
Ah, sorry I should have caught this earlier.

Even with LFS_SHRINKNONRELOCATING enabled, we shouldn't traverse if unless `block_count < lfs->block_count`. `lfs_fs_traverse_` involves a lot of io, and users probably expect `lfs_fs_grow` to be cheap if it is a simple grow operation.

Also I think this can just be `&block_count`? It doesn't look like threshold is used after this.